### ISSUE TITLE
0.7.1: pytest is failing because missing ` jupyter_core.utils.ensure_async`

### ISSUE THREAD

**Issue by kloczek** [jupyter/nbclient#261]
Looks like something is missing or not commites or not finished
```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-nbclient-0.7.1-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-nbclient-0.7.1-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.15, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/tkloczko/rpmbuild/BUILD/nbclient-0.7.1, configfile: pyproject.toml
plugins: asyncio-0.20.2, timeout-2.1.0, flaky-3.7.0
asyncio: mode=auto
collected 0 items / 1 error

================================================================================== ERRORS ==================================================================================
______________________________________________________________________ ERROR collecting test session _______________________________________________________________________
/usr/lib64/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:961: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:961: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:843: in exec_module
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
nbclient/__init__.py:6: in <module>
    from .client import NotebookClient, execute  # noqa: F401
nbclient/client.py:39: in <module>
    from .util import ensure_async, run_hook, run_sync
nbclient/util.py:9: in <module>
    from jupyter_core.utils import ensure_async, run_sync  # noqa: F401
E   ImportError: cannot import name 'ensure_async' from 'jupyter_core.utils' (/usr/lib/python3.8/site-packages/jupyter_core/utils/__init__.py)
============================================================================= warnings summary =============================================================================
../../../../../usr/lib/python3.8/site-packages/jupyter_client/connect.py:27
  /usr/lib/python3.8/site-packages/jupyter_client/connect.py:27: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================================= short test summary info ==========================================================================
ERROR  - ImportError: cannot import name 'ensure_async' from 'jupyter_core.utils' (/usr/lib/python3.8/site-packages/jupyter_core/utils/__init__.py)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================= 1 warning, 1 error in 0.33s ========================================================================
```

**Comment by davidbrochart** [comment]
Do you have [jupyter_core>=4.12,!=~5.0](https://github.com/jupyter/nbclient/blob/645d3bd5a8fb1e8cddc36f6cc4d5ce661a439736/pyproject.toml#L40) installed?

**Comment by kloczek** [comment]
I have 5.1.0.

**Comment by davidbrochart** [comment]
This works:
```console
mamba create -n tmp
mamba activate tmp
mamba install python=3.8
pip install nbclient==0.7.1
pip install pytest xmltodict nbconvert testpath ipykernel ipywidgets
pytest -v ~/mambaforge/envs/tmp/lib/python3.8/site-packages/nbclient/tests
```
Can you provide the output of `pip freeze`?

**Comment by kloczek** [comment]
Here it is
```console
alabaster==0.7.12
appdirs==1.4.4
asttokens==2.1.0
attrs==22.1.0
Babel==2.11.0
backcall==0.2.0
Brlapi==0.8.3
build==0.9.0
charset-normalizer==3.0.1
comm==0.1.0
cssselect==1.1.0
debugpy==1.6.4
decorator==5.1.1
distro==1.8.0
dnspython==2.2.1
docutils==0.19
editables==0.3
entrypoints==0.4
exceptiongroup==1.0.0
executing==1.2.0
extras==1.0.0
fixtures==4.0.0
flaky==3.7.0
gpg===1.17.1-unknown
hatchling==1.11.1
idna==3.4
imagesize==1.4.1
importlib-metadata==5.1.0
iniconfig==1.1.1
ipykernel==6.18.0
ipyparallel==8.4.1
ipython==8.6.0
jedi==0.18.2
Jinja2==3.1.2
jupyter_client==7.4.7
jupyter_core==5.1.0
libcomps==0.1.19
lit==15.0.4
louis==3.23.0
lxml==4.9.1
markdown-it-py==2.1.0
MarkupSafe==2.1.1
matplotlib-inline==0.1.5
mdit-py-plugins==0.3.1
mdurl==0.1.2
meson==0.64.0
myst-parser==0.18.1
nest-asyncio==1.5.6
numpy==1.23.1
packaging==21.3
parso==0.8.3
pathspec==0.10.2
pbr==5.9.0
pep517==0.13.0
pexpect==4.8.0
pickleshare==0.7.5
platformdirs==2.5.2
pluggy==1.0.0
prompt-toolkit==3.0.29
psutil==5.9.2
ptyprocess==0.7.0
pure-eval==0.2.2
Pygments==2.13.0
PyGObject==3.42.2
pyparsing==3.0.9
pytest==7.2.0
pytest-asyncio==0.20.2
pytest-timeout==2.1.0
python-dateutil==2.8.2
pytz==2022.4
PyYAML==6.0
pyzmq==24.0.0
requests==2.28.1
rpm==4.17.0
scour==0.38.2
six==1.16.0
snowballstemmer==2.2.0
Sphinx==5.3.0
sphinxcontrib-applehelp==1.0.2.dev20220730
sphinxcontrib-devhelp==1.0.2.dev20220730
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1.dev20220730
sphinxcontrib-qthelp==1.0.3.dev20220730
sphinxcontrib-serializinghtml==1.1.5
sphinxcontrib_github_alt==1.2
stack-data==0.5.1
testtools==2.5.0
tomli==2.0.1
tornado==6.2
tqdm==4.64.1
traitlets==5.4.0
typing_extensions==4.4.0
urllib3==1.26.12
wcwidth==0.2.5
zipp==3.11.0
```

**Comment by davidbrochart** [comment]
I don't see `nbclient`.

**Comment by kloczek** [comment]
OK one sec ..

**Comment by kloczek** [comment]
Here is result
<details>

```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-nbclient-0.7.1-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-nbclient-0.7.1-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.15, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/tkloczko/rpmbuild/BUILD/nbclient-0.7.1, configfile: pyproject.toml
plugins: asyncio-0.20.2
asyncio: mode=auto
collected 93 items

nbclient/tests/test_client.py ......F.FF.............................................................................. [ 94%]
nbclient/tests/test_util.py .....                                        [100%]

=================================== FAILURES ===================================
________________ test_run_all_notebooks[Interrupt.ipynb-opts6] _________________

input_name = 'Interrupt.ipynb'
opts = {'allow_errors': True, 'interrupt_on_timeout': True, 'kernel_name': 'python', 'timeout': 1}

    @pytest.mark.parametrize(
        ["input_name", "opts"],
        [
            ("Other Comms.ipynb", dict(kernel_name="python")),
            ("Clear Output.ipynb", dict(kernel_name="python")),
            ("Empty Cell.ipynb", dict(kernel_name="python")),
            ("Factorials.ipynb", dict(kernel_name="python")),
            ("HelloWorld.ipynb", dict(kernel_name="python")),
            ("Inline Image.ipynb", dict(kernel_name="python")),
            (
                "Interrupt.ipynb",
                dict(kernel_name="python", timeout=1, interrupt_on_timeout=True, allow_errors=True),
            ),
            ("JupyterWidgets.ipynb", dict(kernel_name="python")),
            ("Skip Exceptions with Cell Tags.ipynb", dict(kernel_name="python")),
            ("Skip Exceptions.ipynb", dict(kernel_name="python", allow_errors=True)),
            ("Skip Execution with Cell Tag.ipynb", dict(kernel_name="python")),
            ("SVG.ipynb", dict(kernel_name="python")),
            ("Unicode.ipynb", dict(kernel_name="python")),
            ("UnicodePy3.ipynb", dict(kernel_name="python")),
            ("update-display-id.ipynb", dict(kernel_name="python")),
            ("Check History in Memory.ipynb", dict(kernel_name="python")),
        ],
    )
    def test_run_all_notebooks(input_name, opts):
        """Runs a series of test notebooks and compares them to their actual output"""
        input_file = os.path.join(current_dir, 'files', input_name)
        input_nb, output_nb = run_notebook(input_file, opts, notebook_resources())
>       assert_notebooks_equal(input_nb, output_nb)

nbclient/tests/test_client.py:339:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

expected = {'cells': [{'cell_type': 'code', 'execution_count': 1, 'metadata': {'collapsed': False}, 'outputs': [{'ename': 'Keyboa...t_type': 'stream', 'text': 'done\n'}], 'source': 'print("done")'}], 'metadata': {}, 'nbformat': 4, 'nbformat_minor': 0}
actual = {'cells': [{'cell_type': 'code', 'metadata': {'collapsed': False, 'execution': {'iopub.status.busy': '2022-11-29T16:44...ents_lexer': 'ipython3', 'nbconvert_exporter': 'python', 'file_extension': '.py'}}, 'nbformat': 4, 'nbformat_minor': 0}

    def assert_notebooks_equal(expected, actual):
        expected_cells = expected['cells']
        actual_cells = actual['cells']
        assert len(expected_cells) == len(actual_cells)

        for expected_cell, actual_cell in zip(expected_cells, actual_cells):
            # Uncomment these to help debug test failures better
            # from pprint import pprint
            # pprint(expected_cell)
            # pprint(actual_cell)
            expected_outputs = expected_cell.get('outputs', [])
            actual_outputs = actual_cell.get('outputs', [])
            normalized_expected_outputs = list(map(normalize_output, expected_outputs))
            normalized_actual_outputs = list(map(normalize_output, actual_outputs))
>           assert normalized_expected_outputs == normalized_actual_outputs
E           AssertionError: assert [{'ename': 'K...nterrupt: ']}] == [{'ename': 'K...nterrupt: ']}]
E             At index 0 diff: {'ename': 'KeyboardInterrupt', 'evalue': '', 'output_type': 'error', 'traceback': ['---------------------------------------------------------------------------', 'KeyboardInterrupt                         Traceback (most recent call last)', '<IPY-INPUT>\n----> 1 while True: continue\n', 'KeyboardInterrupt: ']} != {'output_type': 'error', 'ename': 'KeyboardInterrupt', 'evalue': '', 'traceback': ['---------------------------------------------------------------------------', 'KeyboardInterrupt                         Traceback (most recent call last)', 'Ce...
E
E             ...Full output truncated (2 lines hidden), use '-vv' to show

nbclient/tests/test_client.py:286: AssertionError
------------------------------ Captured log call -------------------------------
ERROR    traitlets:client.py:841 Timeout waiting for execute reply (1s).
ERROR    traitlets:client.py:843 Interrupting kernel
______ test_run_all_notebooks[Skip Exceptions with Cell Tags.ipynb-opts8] ______

input_name = 'Skip Exceptions with Cell Tags.ipynb'
opts = {'kernel_name': 'python'}

    @pytest.mark.parametrize(
        ["input_name", "opts"],
        [
            ("Other Comms.ipynb", dict(kernel_name="python")),
            ("Clear Output.ipynb", dict(kernel_name="python")),
            ("Empty Cell.ipynb", dict(kernel_name="python")),
            ("Factorials.ipynb", dict(kernel_name="python")),
            ("HelloWorld.ipynb", dict(kernel_name="python")),
            ("Inline Image.ipynb", dict(kernel_name="python")),
            (
                "Interrupt.ipynb",
                dict(kernel_name="python", timeout=1, interrupt_on_timeout=True, allow_errors=True),
            ),
            ("JupyterWidgets.ipynb", dict(kernel_name="python")),
            ("Skip Exceptions with Cell Tags.ipynb", dict(kernel_name="python")),
            ("Skip Exceptions.ipynb", dict(kernel_name="python", allow_errors=True)),
            ("Skip Execution with Cell Tag.ipynb", dict(kernel_name="python")),
            ("SVG.ipynb", dict(kernel_name="python")),
            ("Unicode.ipynb", dict(kernel_name="python")),
            ("UnicodePy3.ipynb", dict(kernel_name="python")),
            ("update-display-id.ipynb", dict(kernel_name="python")),
            ("Check History in Memory.ipynb", dict(kernel_name="python")),
        ],
    )
    def test_run_all_notebooks(input_name, opts):
        """Runs a series of test notebooks and compares them to their actual output"""
        input_file = os.path.join(current_dir, 'files', input_name)
        input_nb, output_nb = run_notebook(input_file, opts, notebook_resources())
>       assert_notebooks_equal(input_nb, output_nb)

nbclient/tests/test_client.py:339:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

expected = {'cells': [{'cell_type': 'code', 'execution_count': 1, 'metadata': {'tags': ['raises-exception']}, 'outputs': [{'ename...utput_type': 'stream', 'text': 'ok\n'}], 'source': "print('ok')"}], 'metadata': {}, 'nbformat': 4, 'nbformat_minor': 1}
actual = {'cells': [{'cell_type': 'code', 'metadata': {'tags': ['raises-exception'], 'execution': {'iopub.status.busy': '2022-1...ents_lexer': 'ipython3', 'nbconvert_exporter': 'python', 'file_extension': '.py'}}, 'nbformat': 4, 'nbformat_minor': 1}

    def assert_notebooks_equal(expected, actual):
        expected_cells = expected['cells']
        actual_cells = actual['cells']
        assert len(expected_cells) == len(actual_cells)

        for expected_cell, actual_cell in zip(expected_cells, actual_cells):
            # Uncomment these to help debug test failures better
            # from pprint import pprint
            # pprint(expected_cell)
            # pprint(actual_cell)
            expected_outputs = expected_cell.get('outputs', [])
            actual_outputs = actual_cell.get('outputs', [])
            normalized_expected_outputs = list(map(normalize_output, expected_outputs))
            normalized_actual_outputs = list(map(normalize_output, actual_outputs))
>           assert normalized_expected_outputs == normalized_actual_outputs
E           assert [{'ename': 'E...n: message']}] == [{'ename': 'E...n: message']}]
E             At index 0 diff: {'ename': 'Exception', 'evalue': 'message', 'output_type': 'error', 'traceback': ['---------------------------------------------------------------------------', 'Exception                                 Traceback (most recent call last)', '<IPY-INPUT>\n      1 # üñîçø∂é\n----> 2 raise Exception("message")\n', 'Exception: message']} != {'output_type': 'error', 'ename': 'Exception', 'evalue': 'message', 'traceback': ['---------------------------------------------------------------------------', 'Exception                                 Traceback (most r...
E
E             ...Full output truncated (2 lines hidden), use '-vv' to show

nbclient/tests/test_client.py:286: AssertionError
_____________ test_run_all_notebooks[Skip Exceptions.ipynb-opts9] ______________

input_name = 'Skip Exceptions.ipynb'
opts = {'allow_errors': True, 'kernel_name': 'python'}

    @pytest.mark.parametrize(
        ["input_name", "opts"],
        [
            ("Other Comms.ipynb", dict(kernel_name="python")),
            ("Clear Output.ipynb", dict(kernel_name="python")),
            ("Empty Cell.ipynb", dict(kernel_name="python")),
            ("Factorials.ipynb", dict(kernel_name="python")),
            ("HelloWorld.ipynb", dict(kernel_name="python")),
            ("Inline Image.ipynb", dict(kernel_name="python")),
            (
                "Interrupt.ipynb",
                dict(kernel_name="python", timeout=1, interrupt_on_timeout=True, allow_errors=True),
            ),
            ("JupyterWidgets.ipynb", dict(kernel_name="python")),
            ("Skip Exceptions with Cell Tags.ipynb", dict(kernel_name="python")),
            ("Skip Exceptions.ipynb", dict(kernel_name="python", allow_errors=True)),
            ("Skip Execution with Cell Tag.ipynb", dict(kernel_name="python")),
            ("SVG.ipynb", dict(kernel_name="python")),
            ("Unicode.ipynb", dict(kernel_name="python")),
            ("UnicodePy3.ipynb", dict(kernel_name="python")),
            ("update-display-id.ipynb", dict(kernel_name="python")),
            ("Check History in Memory.ipynb", dict(kernel_name="python")),
        ],
    )
    def test_run_all_notebooks(input_name, opts):
        """Runs a series of test notebooks and compares them to their actual output"""
        input_file = os.path.join(current_dir, 'files', input_name)
        input_nb, output_nb = run_notebook(input_file, opts, notebook_resources())
>       assert_notebooks_equal(input_nb, output_nb)

nbclient/tests/test_client.py:339:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

expected = {'cells': [{'cell_type': 'code', 'execution_count': 1, 'metadata': {'collapsed': False}, 'outputs': [{'ename': 'Except...utput_type': 'stream', 'text': 'ok\n'}], 'source': "print('ok')"}], 'metadata': {}, 'nbformat': 4, 'nbformat_minor': 0}
actual = {'cells': [{'cell_type': 'code', 'metadata': {'collapsed': False, 'execution': {'iopub.status.busy': '2022-11-29T16:44...ents_lexer': 'ipython3', 'nbconvert_exporter': 'python', 'file_extension': '.py'}}, 'nbformat': 4, 'nbformat_minor': 0}

    def assert_notebooks_equal(expected, actual):
        expected_cells = expected['cells']
        actual_cells = actual['cells']
        assert len(expected_cells) == len(actual_cells)

        for expected_cell, actual_cell in zip(expected_cells, actual_cells):
            # Uncomment these to help debug test failures better
            # from pprint import pprint
            # pprint(expected_cell)
            # pprint(actual_cell)
            expected_outputs = expected_cell.get('outputs', [])
            actual_outputs = actual_cell.get('outputs', [])
            normalized_expected_outputs = list(map(normalize_output, expected_outputs))
            normalized_actual_outputs = list(map(normalize_output, actual_outputs))
>           assert normalized_expected_outputs == normalized_actual_outputs
E           assert [{'ename': 'E...n: message']}] == [{'ename': 'E...n: message']}]
E             At index 0 diff: {'ename': 'Exception', 'evalue': 'message', 'output_type': 'error', 'traceback': ['---------------------------------------------------------------------------', 'Exception                                 Traceback (most recent call last)', '<IPY-INPUT>\n      1 # üñîçø∂é\n----> 2 raise Exception("message")\n', 'Exception: message']} != {'output_type': 'error', 'ename': 'Exception', 'evalue': 'message', 'traceback': ['---------------------------------------------------------------------------', 'Exception                                 Traceback (most r...
E
E             ...Full output truncated (2 lines hidden), use '-vv' to show

nbclient/tests/test_client.py:286: AssertionError
=============================== warnings summary ===============================
../../../../../usr/lib/python3.8/site-packages/jupyter_client/connect.py:27
  /usr/lib/python3.8/site-packages/jupyter_client/connect.py:27: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================= slowest 10 durations =============================
11.84s call     nbclient/tests/test_client.py::TestExecute::test_timeout_func
6.20s call     nbclient/tests/test_client.py::test_many_parallel_notebooks
4.44s call     nbclient/tests/test_client.py::TestExecute::test_kernel_death_after_timeout
2.99s call     nbclient/tests/test_client.py::TestExecute::test_timeout
2.98s call     nbclient/tests/test_client.py::test_run_all_notebooks[Interrupt.ipynb-opts6]
2.44s call     nbclient/tests/test_client.py::test_many_async_parallel_notebooks
2.37s call     nbclient/tests/test_client.py::test_execution_timing
2.27s call     nbclient/tests/test_client.py::TestExecute::test_sync_kernel_manager
2.15s call     nbclient/tests/test_client.py::TestExecute::test_disable_stdin
2.09s call     nbclient/tests/test_client.py::TestExecute::test_kernel_death_during_execution
=========================== short test summary info ============================
FAILED nbclient/tests/test_client.py::test_run_all_notebooks[Interrupt.ipynb-opts6] - AssertionError: assert [{'ename': 'K...nterrupt: ']}] == [{'ename': 'K...nt...
FAILED nbclient/tests/test_client.py::test_run_all_notebooks[Skip Exceptions with Cell Tags.ipynb-opts8] - assert [{'ename': 'E...n: message']}] == [{'ename': 'E...n: message']}]
FAILED nbclient/tests/test_client.py::test_run_all_notebooks[Skip Exceptions.ipynb-opts9] - assert [{'ename': 'E...n: message']}] == [{'ename': 'E...n: message']}]
============== 3 failed, 90 passed, 1 warning in 90.65s (0:01:30) ==============
```
</details>

**Comment by davidbrochart** [comment]
Run pytest with `-vv`.

**Comment by davidbrochart** [comment]
Should be fixed by #264.

**Comment by kloczek** [comment]
pytest passes now celanly with one warning
```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-nbclient-0.7.2-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-nbclient-0.7.2-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.15, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/tkloczko/rpmbuild/BUILD/nbclient-0.7.2, configfile: pyproject.toml
plugins: asyncio-0.20.2
asyncio: mode=auto
collected 93 items

nbclient/tests/test_client.py ........................................................................................ [ 94%]
nbclient/tests/test_util.py .....                                        [100%]

=============================== warnings summary ===============================
../../../../../usr/lib/python3.8/site-packages/jupyter_client/connect.py:27
  /usr/lib/python3.8/site-packages/jupyter_client/connect.py:27: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================= slowest 10 durations =============================
12.58s call     nbclient/tests/test_client.py::TestExecute::test_timeout_func
6.29s call     nbclient/tests/test_client.py::test_many_parallel_notebooks
4.61s call     nbclient/tests/test_client.py::TestExecute::test_kernel_death_after_timeout
4.05s call     nbclient/tests/test_client.py::TestExecute::test_populate_language_info
3.51s call     nbclient/tests/test_client.py::TestExecute::test_sync_kernel_manager
3.42s call     nbclient/tests/test_client.py::TestExecute::test_timeout
2.89s call     nbclient/tests/test_client.py::TestExecute::test_reset_kernel_client
2.88s call     nbclient/tests/test_client.py::test_run_all_notebooks[Interrupt.ipynb-opts6]
2.50s call     nbclient/tests/test_client.py::test_many_async_parallel_notebooks
2.34s call     nbclient/tests/test_client.py::test_execution_timing
================== 93 passed, 1 warning in 100.91s (0:01:40) ===================
```
